File: LeetCode/Java/TrappingRainWater.java
import java.util.Stack;

public class TrappingRainWater {

    int res = 0;
        int n = height.length;

        // Method 1

        /*
            for (int i=0;i<n;i++){
                int l = i, r = i;
                
                for (int j = 0;j<i;j++){
                    if (height[j]>height[l]){
                        l = j;
                    }
                }

                for (int j = i;j<n;j++){
                    if (height[j]>height[r]){
                        r = j;
                    }
                }

                res += Math.min(height[l],height[r])-height[i];

            }

            return res;
        */

        // Method 2
        /*
            int[] left = new int[n];
            int[] right = new int[n];

            for (int i=0;i<n;i++){
                left[i] = height[i] ;
                if (i!=0)
                left[i] = Math.max(left[i], left[i-1]);
            }
            for (int i=n-1;i>=0;i--){
                right[i] = height[i] ;
                if (i!=n-1)
                right[i] = Math.max(right[i], right[i+1]);
                res += Math.min(left[i],right[i]) - height[i];
            }
            return res;
        */


        // Method 3
        Stack<Integer> stk = new Stack<>();
        for (int i=0;i<n;i++){
            while (stk.size()!=0 && height[stk.peek()]<height[i]){
                int ind = stk.pop();
                if (stk.size()!=0){
                    int w = i-stk.peek()-1;
                    int h = Math.min(height[i], height[stk.peek()]) - height[ind];
                    res += w*h;
                }
            }
            stk.push(i);
        }
        return res;
        
    }
    
}

File: LeetCode/Python/trapping_rain_water.py
class Solution:
    def trap(self, nums: List[int]) -> int:
        N = len(nums)
        ans = 0
        for i in range(N):
            l = r = i
            for j in range(i):
                if nums[j]>nums[l]:
                    l = j
            for j in range(i,N):
                if nums[j]>nums[r]:
                    r = j
            ans += min(nums[l], nums[r]) - nums[i]
        return ans

class Solution:
    def trap(self, nums: List[int]) -> int:
        N = len(nums)
        ans = 0
        stk = []
        for i in range(N):
            while stk and nums[stk[-1]]<nums[i]:
                ind = stk.pop()
                if stk:
                    h = min(nums[i],nums[stk[-1]])-nums[ind]
                    w = i-stk[-1]-1
                    ans += w*h
            stk.append(i)
        return ans

        